src/mgr/PyModules.cc

```diff
@@ -0,0 +1,456 @@
+  }
+}
+
+void PyModules::shutdown()
+{
+  Mutex::Locker locker(lock);
+
+  // Signal modules to drop out of serve()
+  for (auto i : modules) {
+    auto module = i.second;
+    finisher.queue(new C_StdFunction([module](){
+      module->shutdown();
+    }));
+  }
+
+  for (auto &i : serve_threads) {
+    lock.Unlock();
+    i.second->join();
+    lock.Lock();
+    delete i.second;
+  }
+  serve_threads.clear();
+
+  // Tear down modules
+  for (auto i : modules) {
+    delete i.second;
+  }
+  modules.clear();
+
+  Py_Finalize();
+}
+
+void PyModules::notify_all(const std::string &notify_type,
+                     const std::string &notify_id)
+{
+  Mutex::Locker l(lock);
+
+  dout(10) << __func__ << ": notify_all " << notify_type << dendl;
+  for (auto i : modules) {
+    auto module = i.second;
+    // Send all python calls down a Finisher to avoid blocking
+    // C++ code, and avoid any potential lock cycles.
+    finisher.queue(new C_StdFunction([module, notify_type, notify_id](){
+      module->notify(notify_type, notify_id);
+    }));
+  }
+}
+
+bool PyModules::get_config(const std::string &handle,
+    const std::string &key, std::string *val) const
+{
+  PyThreadState *tstate = PyEval_SaveThread();
+  Mutex::Locker l(lock);
+  PyEval_RestoreThread(tstate);
+
+  const std::string global_key = config_prefix + handle + "." + key;
+
+  if (config_cache.count(global_key)) {
+    *val = config_cache.at(global_key);
+    return true;
+  } else {
+    return false;
+  }
+}
+
+void PyModules::set_config(const std::string &handle,
+    const std::string &key, const std::string &val)
+{
+  const std::string global_key = config_prefix + handle + "." + key;
+
+  Command set_cmd;
+  {
+    PyThreadState *tstate = PyEval_SaveThread();
+    Mutex::Locker l(lock);
+    PyEval_RestoreThread(tstate);
+    config_cache[global_key] = val;
+
+    std::ostringstream cmd_json;
+
+    JSONFormatter jf;
+    jf.open_object_section("cmd");
+    jf.dump_string("prefix", "config-key put");
+    jf.dump_string("key", global_key);
+    jf.dump_string("val", val);
+    jf.close_section();
+    jf.flush(cmd_json);
+
+    set_cmd.run(&monc, cmd_json.str());
+  }
+  set_cmd.wait();
+
+  // FIXME: is config-key put ever allowed to fail?
+  assert(set_cmd.r == 0);
+}
+
+std::vector<ModuleCommand> PyModules::get_commands()
+{
+  Mutex::Locker l(lock);
+
+  std::vector<ModuleCommand> result;
+  for (auto i : modules) {
+    auto module = i.second;
+    auto mod_commands = module->get_commands();
+    for (auto j : mod_commands) {
+      result.push_back(j);
+    }
+  }
+
+  return result;
+}
+
+void PyModules::insert_config(const std::map<std::string,
+                              std::string> &new_config)
+{
+  Mutex::Locker l(lock);
+
+  dout(4) << "Loaded " << new_config.size() << " config settings" << dendl;
+  config_cache = new_config;
+}
+
+void PyModules::log(const std::string &handle,
+    unsigned level, const std::string &record)
```

Comment:
With level being unsigned, I get a build failure:

````
/home/tserong/src/github/SUSE/ceph/src/mgr/PyModules.cc: In member function ‘void PyModules::log(const string&, unsigned int, const string&)’:
/home/tserong/src/github/SUSE/ceph/src/common/dout.h:57:24: warning: comparison between signed and unsigned integer expressions [-Wsign-compare]
       char __array[((v >= -1) && (v <= 200)) ? 0 : -1] __attribute__((unused)); \
                         
/home/tserong/src/github/SUSE/ceph/src/common/dout.h:57:24: note: in definition of macro ‘dout_impl’
       char __array[((v >= -1) && (v <= 200)) ? 0 : -1] __attribute__((unused)); \
                        ^~
/home/tserong/src/github/SUSE/ceph/src/common/debug.h:23:17: note: in expansion of macro ‘ldout’
 #define dout(v) ldout((g_ceph_context), v)
                 ^~~~~
/home/tserong/src/github/SUSE/ceph/src/mgr/PyModules.cc:453:3: note: in expansion of macro ‘dout’
   dout(level) << record << dendl;
   ^
/home/tserong/src/github/SUSE/ceph/src/common/dout.h:57:78: error: size of array ‘__array’ is negative
       char __array[((v >= -1) && (v <= 200)) ? 0 : -1] __attribute__((unused)); \
                                                                              ^
/home/tserong/src/github/SUSE/ceph/src/common/dout.h:66:24: note: in expansion of macro ‘dout_impl’
 #define ldout(cct, v)  dout_impl(cct, dout_subsys, v) dout_prefix
                        ^~~~~~~~~
/home/tserong/src/github/SUSE/ceph/src/common/debug.h:23:17: note: in expansion of macro ‘ldout’
 #define dout(v) ldout((g_ceph_context), v)
                 ^~~~~
/home/tserong/src/github/SUSE/ceph/src/mgr/PyModules.cc:453:3: note: in expansion of macro ‘dout’
   dout(level) << record << dendl;
   ^
````

Changing level to an int here (and also in PyModules.h) fixes it.